File: benchmarks/typescript/load_feed.ts
type Feed = {
  user_id: string;
  posts: string[];
};

type FeedError =
  | { kind: "network"; status: number }
  | { kind: "decode"; reason: string };

type FeedResult =
  | { ok: true; value: Feed }
  | { ok: false; error: FeedError };

export async function loadFeed(userId: string): Promise<FeedResult> {
  try {
    const userResp = await fetch(`/api/users/${userId}`);
    if (!userResp.ok) {
      return { ok: false, error: { kind: "network", status: userResp.status } };
    }
    const postsResp = await fetch(`/api/users/${userId}/posts`);
    if (!postsResp.ok) {
      return { ok: false, error: { kind: "network", status: postsResp.status } };
    }
    const postsData = (await postsResp.json()) as { titles: string[] };
    return {
      ok: true,
      value: { user_id: userId, posts: postsData.titles },
    };
  } catch (e) {
    return { ok: false, error: { kind: "decode", reason: String(e) } };
  }
}

File: benchmarks/go/load_feed.go
package loadfeed

import (
	"encoding/json"
	"fmt"
	"net/http"
)

type Feed struct {
	UserID string
	Posts  []string
}

type FeedError struct {
	Kind   string // "network" or "decode"
	Status int
	Reason string
}

func (e FeedError) Error() string {
	if e.Kind == "network" {
		return fmt.Sprintf("network error: status %d", e.Status)
	}
	return "decode error: " + e.Reason
}

func LoadFeed(client *http.Client, userID string) (Feed, error) {
	userResp, err := client.Get(fmt.Sprintf("/api/users/%s", userID))
	if err != nil {
		return Feed{}, FeedError{Kind: "decode", Reason: err.Error()}
	}
	defer userResp.Body.Close()
	if userResp.StatusCode >= 400 {
		return Feed{}, FeedError{Kind: "network", Status: userResp.StatusCode}
	}

	postsResp, err := client.Get(fmt.Sprintf("/api/users/%s/posts", userID))
	if err != nil {
		return Feed{}, FeedError{Kind: "decode", Reason: err.Error()}
	}
	defer postsResp.Body.Close()
	if postsResp.StatusCode >= 400 {
		return Feed{}, FeedError{Kind: "network", Status: postsResp.StatusCode}
	}

	var postsData struct {
		Titles []string `json:"titles"`
	}
	if err := json.NewDecoder(postsResp.Body).Decode(&postsData); err != nil {
		return Feed{}, FeedError{Kind: "decode", Reason: err.Error()}
	}
	return Feed{UserID: userID, Posts: postsData.Titles}, nil
}
